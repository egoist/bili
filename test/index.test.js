import path from 'path'
import Bili from '../src'

process.env.BABEL_ENV = 'anything-not-test'

function fixture(...args) {
  return path.join(__dirname, 'fixtures', ...args)
}

function generate(options) {
  return Bili.generate({
    config: false,
    ...options
  })
}

function snapshot({ title, input, ...args }) {
  input = Array.isArray(input) ? input : [input]
  test(title, async () => {
    const { bundles } = await generate({
      input,
      ...args
    })
    expect(Object.keys(bundles)
      .sort()
      .map(filepath => [bundles[filepath].relative, bundles[filepath].code])).toMatchSnapshot()
  })
}

snapshot({
  title: 'defaults',
  input: 'index.js',
  cwd: fixture('defaults')
})

snapshot({
  title: 'buble:async',
  input: 'async.js',
  js: 'buble',
  cwd: fixture('buble')
})

snapshot({
  title: 'buble:async-and-object-rest-spread',
  input: 'async-dot-dot-dot.js',
  js: 'buble',
  cwd: fixture('buble')
})

snapshot({
  title: 'buble:react-jsx',
  input: 'react-jsx.js',
  js: 'buble',
  cwd: fixture('buble')
})

snapshot({
  title: 'buble:vue-jsx',
  input: 'vue-jsx.js',
  jsx: 'vue',
  js: 'buble',
  cwd: fixture('buble')
})

snapshot({
  title: 'banner:true',
  banner: true,
  input: 'index.js',
  cwd: fixture('banner')
})

snapshot({
  title: 'banner:object',
  input: 'default.js',
  banner: {
    year: '2018',
    author: 'author',
    license: 'GPL',
    name: 'name',
    version: '1.2.3'
  },
  cwd: fixture()
})

snapshot({
  title: 'no-js-transform',
  js: false,
  input: 'index.js',
  cwd: fixture('no-js-transform')
})

snapshot({
  title: 'banner:string',
  input: 'default.js',
  banner: 'woot',
  cwd: fixture()
})

snapshot({
  title: 'exclude file',
  input: 'index.js',
  external: ['./foo.js'],
  cwd: fixture('exclude-file')
})

snapshot({
  title: 'extendOptions',
  format: 'umd,umd-min,cjs',
  input: ['foo.js', 'bar.js'],
  cwd: fixture('extend-options'),
  extendOptions(options, { input, format, compress }) {
    if (input.endsWith('foo.js')) {
      options.moduleName = 'endsWithFoo'
    }
    if (format === 'umd') {
      options.moduleName = 'umd'
    }
    if (compress) {
      options.moduleName = 'min'
    }
    return options
  }
})

snapshot({
  title: 'inline:true',
  inline: true,
  input: 'index.js',
  cwd: fixture('inline')
})

snapshot({
  title: 'async',
  input: 'index.js',
  cwd: fixture('async')
})

describe('multi formats without suffix error', () => {
  test('it throws', async () => {
    expect.assertions(1)
    try {
      await generate({
        input: 'index.js',
        format: ['cjs', 'umd'],
        filename: '[name].js',
        cwd: fixture('defaults'),
        moduleName: 'foo'
      })
    } catch (err) {
      expect(err.message).toMatch(/Multiple files are emitting to the same path/)
    }
  })

  test('it does not throw', async () => {
    await generate({
      input: 'index.js',
      format: ['umd-min', 'umd'],
      filename: '[name].js',
      cwd: fixture('defaults'),
      moduleName: 'foo'
    })
  })
})

test('cwd', async () => {
  // This tests two things
  const bili = await generate({
    // 1. resolve `input` from `cwd`
    input: 'index.js',
    cwd: fixture('defaults')
  })
  // 2. output file is relative to `cwd`
  const outputPath = Object.keys(bili.bundles)[0]
  expect(outputPath).toMatch('/test/fixtures/defaults/dist/index.cjs.js')
})

snapshot({
  title: 'virtualModules',
  input: './dayyum',
  cwd: fixture('virtual'),
  virtualModules: {
    './dayyum': `import bar from './bar'\nexport default bar + 123`,
    // Relative path is relative to process.cwd()
    './bar': `import baz from 'baz';export default baz + 1`,
    baz: `export default 2`
  }
})

snapshot({
  title: 'target:node',
  input: 'index.js',
  cwd: fixture('target/node'),
  target: 'node'
})

snapshot({
  title: 'babel:with-config',
  input: 'index.js',
  cwd: fixture('babel/with-config')
})

snapshot({
  title: 'babel:disable-config',
  input: 'index.js',
  cwd: fixture('babel/with-config'),
  babel: {
    babelrc: false
  }
})

snapshot({
  title: 'babel:object-rest-spread',
  input: 'index.js',
  cwd: fixture('babel/object-rest-spread')
})

snapshot({
  title: 'uglify',
  input: 'index.js',
  cwd: fixture('uglify'),
  format: 'cjs-min'
})

snapshot({
  title: 'prettier (should ignore min files)',
  input: 'index.js',
  cwd: fixture('prettier'),
  format: 'cjs-min',
  pretty: true
})

snapshot({
  title: 'prettier (should set options)',
  input: 'index.js',
  cwd: fixture('prettier'),
  format: 'cjs',
  pretty: true,
  prettier: {
    tabWidth: 4,
    singleQuote: false,
    trailingComma: 'all',
    semi: false
  }
})

snapshot({
  title: 'inline-certain-modules',
  input: 'index.js',
  cwd: fixture('inline-certain-modules'),
  inline: 'fake-module'
})
