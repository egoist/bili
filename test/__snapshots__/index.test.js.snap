// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async 1`] = `
Array [
  Array [
    "test/fixtures/async/dist/index.cjs.js",
    "'use strict';

var _this = undefined;

var a = {
  a: 1
};
var index = (function () {
  return new Promise(function ($return, $error) {
    return $return(Object.assign({}, a, {
      b: _this.b
    }));
  });
});

module.exports = index;
",
  ],
]
`;

exports[`babel:disable-config 1`] = `
Array [
  Array [
    "test/fixtures/babel/with-config/dist/index.cjs.js",
    "'use strict';

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError(\\"Cannot call a class as a function\\");
  }
}

var _default = function _default() {
  _classCallCheck(this, _default);
};

module.exports = _default;
",
  ],
]
`;

exports[`babel:object-rest-spread 1`] = `
Array [
  Array [
    "test/fixtures/babel/object-rest-spread/dist/index.cjs.js",
    "'use strict';

var index = Object.assign({}, a);

module.exports = index;
",
  ],
]
`;

exports[`babel:with-config 1`] = `
Array [
  Array [
    "test/fixtures/babel/with-config/dist/index.cjs.js",
    "'use strict';

class index {}

module.exports = index;
",
  ],
]
`;

exports[`banner:object 1`] = `
Array [
  Array [
    "test/fixtures/dist/default.cjs.js",
    "/*!
 * name v1.2.3
 * (c) 2018-present author
 * Released under the GPL License.
 */
'use strict';

var _default = {};

module.exports = _default;
",
  ],
]
`;

exports[`banner:string 1`] = `
Array [
  Array [
    "test/fixtures/dist/default.cjs.js",
    "woot
'use strict';

var _default = {};

module.exports = _default;
",
  ],
]
`;

exports[`banner:true 1`] = `
Array [
  Array [
    "test/fixtures/banner/dist/foo.cjs.js",
    "/*!
 * foo v0.0.0
 * (c) 2015-present name <email@email.com>
 * Released under the MIT License.
 */
'use strict';

var index = 42;

module.exports = index;
",
  ],
]
`;

exports[`buble:async 1`] = `
Array [
  Array [
    "test/fixtures/buble/dist/async.cjs.js",
    "'use strict';

var a = function () { return new Promise(function ($return, $error) {
  return $return({
    a: 'a'
  });
}); };

var async = a().then(function (res) { return Object.assign({}, res); });

module.exports = async;
",
  ],
]
`;

exports[`buble:async-and-object-rest-spread 1`] = `
Array [
  Array [
    "test/fixtures/buble/dist/async-dot-dot-dot.cjs.js",
    "'use strict';

function asyncDotDotDot () {
  return new Promise(function ($return, $error) {
    return Promise.resolve(this.bar()).then(function ($await_1) {
      try {
        return $return(Object.assign({}, $await_1));
      } catch ($boundEx) {
        return $error($boundEx);
      }
    }, $error);
  }.bind(this));
}

module.exports = asyncDotDotDot;
",
  ],
]
`;

exports[`buble:react-jsx 1`] = `
Array [
  Array [
    "test/fixtures/buble/dist/react-jsx.cjs.js",
    "'use strict';

var reactJsx = React.createElement(\\"div\\", {
  a: 1
}, \\"hi\\");

module.exports = reactJsx;
",
  ],
]
`;

exports[`buble:vue-jsx 1`] = `
Array [
  Array [
    "test/fixtures/buble/dist/vue-jsx.cjs.js",
    "'use strict';

var vueJsx = {
  render: function render() {
    var h = arguments[0];
    return h(\\"div\\", {
      attrs: {
        a: 1
      }
    }, [\\"hi\\"]);
  }

};

module.exports = vueJsx;
",
  ],
]
`;

exports[`defaults 1`] = `
Array [
  Array [
    "test/fixtures/defaults/dist/index.cjs.js",
    "'use strict';

var index = 42;

module.exports = index;
",
  ],
]
`;

exports[`exclude file 1`] = `
Array [
  Array [
    "test/fixtures/exclude-file/dist/index.cjs.js",
    "'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var foo = _interopDefault(require('./foo.js'));



module.exports = foo;
",
  ],
]
`;

exports[`extendOptions 1`] = `
Array [
  Array [
    "test/fixtures/extend-options/dist/bar.cjs.js",
    "'use strict';

var bar = 'bar';

module.exports = bar;
",
  ],
  Array [
    "test/fixtures/extend-options/dist/bar.js",
    "(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.umd = factory());
}(this, (function () { 'use strict';

	var bar = 'bar';

	return bar;

})));
",
  ],
  Array [
    "test/fixtures/extend-options/dist/bar.min.js",
    "!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):e.min=n()}(this,function(){\\"use strict\\";return\\"bar\\"});
",
  ],
  Array [
    "test/fixtures/extend-options/dist/foo.cjs.js",
    "'use strict';

var foo = 'foo';

module.exports = foo;
",
  ],
  Array [
    "test/fixtures/extend-options/dist/foo.js",
    "(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
	typeof define === 'function' && define.amd ? define(factory) :
	(global.umd = factory());
}(this, (function () { 'use strict';

	var foo = 'foo';

	return foo;

})));
",
  ],
  Array [
    "test/fixtures/extend-options/dist/foo.min.js",
    "!function(e,n){\\"object\\"==typeof exports&&\\"undefined\\"!=typeof module?module.exports=n():\\"function\\"==typeof define&&define.amd?define(n):e.min=n()}(this,function(){\\"use strict\\";return\\"foo\\"});
",
  ],
]
`;

exports[`inline:true 1`] = `
Array [
  Array [
    "test/fixtures/inline/dist/index.cjs.js",
    "'use strict';

var pMap = (iterable, mapper, opts) => new Promise((resolve, reject) => {
	opts = Object.assign({
		concurrency: Infinity
	}, opts);

	if (typeof mapper !== 'function') {
		throw new TypeError('Mapper function is required');
	}

	const concurrency = opts.concurrency;

	if (!(typeof concurrency === 'number' && concurrency >= 1)) {
		throw new TypeError(\`Expected \\\\\`concurrency\\\\\` to be a number from 1 and up, got \\\\\`\${concurrency}\\\\\` (\${typeof concurrency})\`);
	}

	const ret = [];
	const iterator = iterable[Symbol.iterator]();
	let isRejected = false;
	let iterableDone = false;
	let resolvingCount = 0;
	let currentIdx = 0;

	const next = () => {
		if (isRejected) {
			return;
		}

		const nextItem = iterator.next();
		const i = currentIdx;
		currentIdx++;

		if (nextItem.done) {
			iterableDone = true;

			if (resolvingCount === 0) {
				resolve(ret);
			}

			return;
		}

		resolvingCount++;

		Promise.resolve(nextItem.value)
			.then(el => mapper(el, i))
			.then(
				val => {
					ret[i] = val;
					resolvingCount--;
					next();
				},
				err => {
					isRejected = true;
					reject(err);
				}
			);
	};

	for (let i = 0; i < concurrency; i++) {
		next();

		if (iterableDone) {
			break;
		}
	}
});

var pAll = (iterable, opts) => pMap(iterable, el => el(), opts);

var index = {
  all: pAll,
  map: pMap
};

module.exports = index;
",
  ],
]
`;

exports[`inline-certain-modules 1`] = `
Array [
  Array [
    "test/fixtures/inline-certain-modules/dist/index.cjs.js",
    "'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var another = _interopDefault(require('another-module'));

var fakeModule = function (a) {
  return a;
};

var index = fakeModule(1, another);

module.exports = index;
",
  ],
]
`;

exports[`no-js-transform 1`] = `
Array [
  Array [
    "test/fixtures/no-js-transform/dist/index.cjs.js",
    "'use strict';

var index = () => {};

module.exports = index;
",
  ],
]
`;

exports[`prettier (should ignore min files) 1`] = `
Array [
  Array [
    "test/fixtures/prettier/dist/index.cjs.min.js",
    "\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});var a=1,d=\\"double stringh\\";function done(){return{prop:343,another:782}}exports.a=a,exports.d=d,exports.done=done;
",
  ],
]
`;

exports[`prettier (should set options) 1`] = `
Array [
  Array [
    "test/fixtures/prettier/dist/index.cjs.js",
    "\\"use strict\\"

Object.defineProperty(exports, \\"__esModule\\", { value: true })

var a = 1
var d = \\"double stringh\\"
function done() {
    var t = {
        prop: 343,
        another: 782,
    }
    return t
}

exports.a = a
exports.d = d
exports.done = done
",
  ],
]
`;

exports[`target:node 1`] = `
Array [
  Array [
    "test/fixtures/target/node/dist/index.cjs.js",
    "'use strict';

class Foo {}

module.exports = Foo;
",
  ],
]
`;

exports[`uglify 1`] = `
Array [
  Array [
    "test/fixtures/uglify/dist/index.cjs.min.js",
    "\\"use strict\\";Object.defineProperty(exports,\\"__esModule\\",{value:!0});var a=1;exports.a=a;
",
  ],
]
`;

exports[`virtualModules 1`] = `
Array [
  Array [
    "test/fixtures/virtual/dist/dayyum.cjs.js",
    "'use strict';

var baz = 2;

var bar = baz + 1;

var dayyum = bar + 123;

module.exports = dayyum;
",
  ],
]
`;
